### PR TITLE
Packet network buffer

### PR DESCRIPTION
**Pull request by TheMode**

### CODE REVIEW

File: src/main/java/net/minestom/server/crypto/ArgumentSignatures.java [hunk 1]
```diff
@@ -1,34 +1,34 @@
...
     @Override
-    public void write(@NotNull BinaryWriter writer) {
-        writer.writeVarIntList(entries, BinaryWriter::write);
+    public void write(@NotNull NetworkBuffer writer) {
+        writer.writeCollection(entries);
     }
 
-    public record Entry(@NotNull String name, @NotNull MessageSignature signature) implements Writeable {
-        public Entry(BinaryReader reader) {
-            this(reader.readSizedString(), new MessageSignature(reader));
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/crypto/ArgumentSignatures.java:26]: Doesn't the indirection via Type reduce performance?

File: src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java [hunk 2]
```diff
@@ -32,21 +33,21 @@ public record HandshakePacket(int protocolVersion, @NotNull String serverAddress
     private static final Component INVALID_VERSION_TEXT = Component.text("Invalid Version, please use " + MinecraftServer.VERSION_NAME, NamedTextColor.RED);
     private static final Component INVALID_BUNGEE_FORWARDING = Component.text("If you wish to use IP forwarding, please enable it in your BungeeCord config as well!", NamedTextColor.RED);
 
-    public HandshakePacket(BinaryReader reader) {
-        this(reader.readVarInt(), reader.readSizedString(BungeeCordProxy.isEnabled() ? Short.MAX_VALUE : 255),
-                reader.readUnsignedShort(), reader.readVarInt());
+    public HandshakePacket(NetworkBuffer reader) {
+        this(reader.read(VAR_INT), reader.read(STRING),
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java:37]: Lost the max-size parameter, is that intended?

File: src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java [hunk 3]
```diff
@@ -32,21 +33,21 @@ public record HandshakePacket(int protocolVersion, @NotNull String serverAddress
...
+                reader.read(SHORT), reader.read(VAR_INT));
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java:38]: Short and Unsigned Short are not the same

File: src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java [hunk 4]
```diff
@@ -32,21 +33,21 @@ public record HandshakePacket(int protocolVersion, @NotNull String serverAddress
...
+        writer.write(VAR_INT, protocolVersion);
         int maxLength = BungeeCordProxy.isEnabled() ? Short.MAX_VALUE : 255;
         if (serverAddress.length() > maxLength) {
             throw new IllegalArgumentException("serverAddress is " + serverAddress.length() + " characters long, maximum allowed is " + maxLength);
         }
-        writer.writeSizedString(serverAddress);
-        writer.writeUnsignedShort(serverPort);
-        writer.writeVarInt(nextState);
+        writer.write(STRING, serverAddress);
+        writer.write(SHORT, (short) serverPort);
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java:49]: Unsigned short != Short

File: src/main/java/net/minestom/server/network/packet/client/play/ClientAdvancementTabPacket.java [hunk 5]
```diff
@@ -1,37 +1,39 @@
...
     private ClientAdvancementTabPacket(ClientAdvancementTabPacket packet) {
         this(packet.action, packet.tabIdentifier);
     }
 
-    private static ClientAdvancementTabPacket read(BinaryReader reader) {
-        var action = AdvancementAction.values()[reader.readVarInt()];
-        var tabIdentifier = action == AdvancementAction.OPENED_TAB ? reader.readSizedString(256) : null;
+    private static ClientAdvancementTabPacket read(@NotNull NetworkBuffer reader) {
+        var action = AdvancementAction.values()[reader.read(VAR_INT)];
+        var tabIdentifier = action == AdvancementAction.OPENED_TAB ? reader.read(STRING) : null;
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/client/play/ClientAdvancementTabPacket.java:24]: Size limit lost

File: src/main/java/net/minestom/server/network/packet/client/play/ClientChatMessagePacket.java [hunk 6]
```diff
@@ -17,20 +18,20 @@ public record ClientChatMessagePacket(@NotNull String message,
         }
     }
 
-    public ClientChatMessagePacket(BinaryReader reader) {
-        this(reader.readSizedString(256),
-                reader.readLong(), reader.readLong(), new MessageSignature(reader),
-                reader.readBoolean(),
+    public ClientChatMessagePacket(NetworkBuffer reader) {
+        this(reader.read(STRING),
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/client/play/ClientChatMessagePacket.java:22]: Lost size limit

File: src/main/java/net/minestom/server/network/packet/client/play/ClientChatPreviewPacket.java [hunk 7]
```diff
@@ -1,18 +1,20 @@
...
 import org.jetbrains.annotations.NotNull;
 
+import static net.minestom.server.network.NetworkBuffer.INT;
+import static net.minestom.server.network.NetworkBuffer.STRING;
+
 public record ClientChatPreviewPacket(int queryId, @NotNull String query) implements ClientPacket {
-    public ClientChatPreviewPacket(BinaryReader reader) {
-        this(reader.readInt(), reader.readSizedString(256));
+    public ClientChatPreviewPacket(NetworkBuffer reader) {
+        this(reader.read(INT), reader.read(STRING));
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/client/play/ClientChatPreviewPacket.java:12]: Lost size limit

File: src/main/java/net/minestom/server/network/packet/server/play/data/LightData.java [hunk 8]
```diff
@@ -1,35 +1,36 @@
...
-                BitSet.valueOf(reader.readLongArray()), BitSet.valueOf(reader.readLongArray()),
-                BitSet.valueOf(reader.readLongArray()), BitSet.valueOf(reader.readLongArray()),
-                reader.readVarIntList(r -> r.readBytes(r.readVarInt())), reader.readVarIntList(r -> r.readBytes(r.readVarInt())));
+                        @NotNull List<byte[]> skyLight,
+                        @NotNull List<byte[]> blockLight) implements NetworkBuffer.Writer {
+    public LightData(@NotNull NetworkBuffer reader) {
+        this(reader.read(BOOLEAN),
+                BitSet.valueOf(reader.read(LONG_ARRAY)), BitSet.valueOf(reader.read(LONG_ARRAY)),
+                BitSet.valueOf(reader.read(LONG_ARRAY)), BitSet.valueOf(reader.read(LONG_ARRAY)),
+                reader.readCollection(BYTE_ARRAY), reader.readCollection(BYTE_ARRAY));
```
Comment (jglrxavpok) [inline src/main/java/net/minestom/server/network/packet/server/play/data/LightData.java:20]: I find the new version easier to understand.

**Review by jglrxavpok — COMMENTED** [review]
The size limit has disappearing, and unsigned short no longer has support with this iteration (we store them in ints because Java has no unsigned types).
Other than that, I did not see any other flaw in the PR.

File: src/main/java/net/minestom/server/network/packet/client/play/ClientSetBeaconEffectPacket.java [hunk 9]
```diff
@@ -1,24 +1,26 @@
...
-    public void write(@NotNull BinaryWriter writer) {
-        writer.writeBoolean(primaryEffect != null);
-        if (primaryEffect != null) writer.writeVarInt(primaryEffect.id());
-        writer.writeBoolean(secondaryEffect != null);
-        if (secondaryEffect != null) writer.writeVarInt(secondaryEffect.id());
+    public void write(@NotNull NetworkBuffer writer) {
+        writer.write(BOOLEAN, primaryEffect != null);
+        if (primaryEffect != null) writer.write(VAR_INT, primaryEffect.id());
+        writer.write(BOOLEAN, secondaryEffect != null);
+        if (secondaryEffect != null) writer.write(VAR_INT, secondaryEffect.id());
```
[the remark this answers is not in the document]
Comment (TheMode) [reply src/main/java/net/minestom/server/network/packet/client/play/ClientSetBeaconEffectPacket.java:24]: Not really, as `#id()` could throw an NPE

File: src/main/java/net/minestom/server/network/packet/server/play/CustomChatCompletionPacket.java [hunk 10]
```diff
@@ -1,27 +1,28 @@
...
 public record CustomChatCompletionPacket(@NotNull Action action,
                                          @NotNull List<@NotNull String> entries) implements ServerPacket {
     public CustomChatCompletionPacket {
         entries = List.copyOf(entries);
     }
 
-    public CustomChatCompletionPacket(BinaryReader reader) {
-        this(Action.values()[reader.readVarInt()], reader.readVarIntList(BinaryReader::readSizedString));
+    public CustomChatCompletionPacket(@NotNull NetworkBuffer reader) {
+        this(reader.readEnum(Action.class), reader.readCollection(STRING));
```
[the remark this answers is not in the document]
Comment (TheMode) [reply src/main/java/net/minestom/server/network/packet/server/play/CustomChatCompletionPacket.java:19]: They all do yes

File: src/main/java/net/minestom/server/crypto/ArgumentSignatures.java (hunk 1 above)
Comment (TheMode) [reply src/main/java/net/minestom/server/crypto/ArgumentSignatures.java:26]: In theory yes, but in practice it is very minimal. And having the ability to describe a writable structure using constants could allow some interesting optimizations.

Though performance isn't the main goal here. This version is much more maintainable.

**Pull request merged**